### PR TITLE
`Notification` to `NameTracker`, and methods to create the relationship

### PR DESCRIPTION
**Pull request by nimmolo**
**Migration**
- Rename the table from `notifications` to `name_trackers`
- Add a record to `interests` corresponding to each record in `name_trackers`

**App**
- Rename the class from `Notification` to `NameTracker`, update calls to `Notification` throughout the app
- Add `NameTracker` to `Interest.all_type_tags` and a [validator](https://stackoverflow.com/questions/40681599/rails-polymorphic-association-restrict-allowed-classes) for `Interest.target_type`
- Have the "Email Tracking" form create, update or delete an `Interest` at the same time as the `NameTracker` with an association between the records
- Make it that deleting an `Interest` targeting a `NameTracker` deletes the `NameTracker`, and vice versa

**Test**
- Add  `Interest` fixtures for `NameTracker`s
- Add tests, adjust expectations

### CODE REVIEW

File: app/controllers/interest_controller.rb [hunk 1]
```diff
@@ -26,25 +26,15 @@ class InterestController < ApplicationController
...
-    interests = find_relevant_interests
-    @targets = notifications + interests
+    @interests = find_relevant_interests
 
     @pages = paginate_numbers(:page, 50)
-    @pages.num_total = @targets.length
-    @targets = @targets[@pages.from..@pages.to]
+    @pages.num_total = @interests.length
+    @interests = @interests[@pages.from..@pages.to]
   end
```
Comment (nimmolo) [inline app/controllers/interest_controller.rb:34]: Switches from `@targets` to more straightforward `@interests` for the view - we no longer have to add notifications to the interests.

File: app/controllers/interest_controller.rb [hunk 2]
```diff
@@ -161,8 +156,8 @@ def redirect_to_target_or_list_interests
```
Comment (nimmolo) [inline app/controllers/interest_controller.rb:156]: Because there is no `NameTracker.show_controller`.

File: app/models/interest.rb [hunk 3]
```diff
@@ -56,14 +56,27 @@ class Interest < AbstractModel
...
+    Name
+    NameTracker
+    Observation
+    Project
+    SpeciesList
+  ].freeze
+
+  validates :target_type, inclusion: { in: ALL_INTEREST_TYPES }
+
   # Returns Array of all models (Classes) which take interests.
```
Comment (nimmolo) [inline app/models/interest.rb:70]: New validation

File: app/views/interest/list_interests.html.erb [hunk 4]
```diff
@@ -1,36 +1,34 @@
...
+            <%= link_to(item.state ? :list_interests_turn_off.t :
+                                      :list_interests_turn_on.t,
+                        controller: :interest, action: :set_interest,
+                        type: item.target_type, id: item.target_id,
+                        state: item.state ? -1 : 1) %> |
+            <%= link_to(:DESTROY.t,
+                        controller: :interest, action: :set_interest,
+                        type: item.target_type, id: item.target_id,
+                        state: 0) %>
           </td>
```
Comment (nimmolo) [inline app/views/interest/list_interests.html.erb:32]: Simplified `:list_interests` just shows interests, now

File: test/controllers/interest_controller_test.rb [hunk 5]
```diff
@@ -88,35 +91,48 @@ def test_set_interest
...
+    coprinus_comatus_tracker = name_trackers(:coprinus_comatus_name_tracker)
+
+    # Now, delete the interest in the name_tracker, and be sure also deleted
+    login("rolf")
+    get(:set_interest, params: { type: "NameTracker",
+                                 id: coprinus_comatus_tracker.id,
+                                 state: 0 })
+    assert_flash_success
     assert_equal(0, Interest.where(user_id: rolf.id).length)
+    assert_equal(0, NameTracker.where(user_id: rolf.id).length)
```
Comment (nimmolo) [inline test/controllers/interest_controller_test.rb:125]: New test for deleting nametracker when interest deleted.

**Comment by nimmolo** [comment]
OK, shall I merge this, and the next one? They're not that complicated.
I kept them separate to facilitate review, but i'm not sure anyone actually wants to review them.
 
Feel free in the future to drop a note that you'll pass on reviewing.
If i'm not confident about the code I'll say so.

**Comment by pellaea** [comment]
Oh man, I'm sorry, I must have forgotten about this PR.  I don't have any
more time to look at PRs now than I have had for the past few days.  And it
doesn't look good for the rest of the week.

**Comment by nimmolo** [comment]
All good. Thank you for the guidance at the outset!
I'm so ready to be done with all these controllers.

**Comment by JoeCohen** [comment]
Nimmo:
Sorry to be so negligent in reviewing your PRs.
If you like, I'll try to find time to review this this afternoon. But I
can't promise to have anything useful to say (and certainly cannot be as
helpful as Jason -- even his cursory reviews.
So I don't have any problem if you don't want to wait on me.
The ones that have been merged and deployed have been AFAIK problem free,
and the code itself is very nice in standardizing things with the bonus of
lots of improvements in readability and metrics.
-- Joe
PS: You beat me to it. That's very helpful as I can now merge this with my
draft PR #1241, and don't have to worry about colliding with you.

**Comment by nimmolo** [comment]
Thanks Joe, there’s another one merging in a minute.

I’ve committed all your changes, in recent comments...
I hope i’m not doing anything that makes you feel they’re unwelcome, they always improve my code and you catch things I don’t!

Anything you have time for is welcome, and if not — please just leave a comment that you’ll pass.

AN

**Pull request merged**